### ISSUE TITLE
Start/resume track stutters

### ISSUE THREAD

**Issue by The-Michael-R** [MusicPlayerDaemon/MPD#420]
Since the last update of mpd on archLinux@RasPi 3B with Hifiberry DAC I experience a small stutter after `mpc play` (or skip to another position in the track) of a track or a stream. It seems the first frame is played twice. After that stutter everything plays fine. Also the no stutter continuously playing through the playlist (going from one track to the next). Only after a a user-interaction (remote or local mpc) shows this issue.

The `--verbose` switch doesn't show any problem during the stutter. But below is the beginning of the log as it might hold lots of information that might be useful:
```
>mpd -V  
Music Player Daemon 0.21.1 (0.21.1)
[...]

> mpd --no-daemon --stdout --verbose
config_file: loading file /etc/mpd.conf
exception: bind to '0.0.0.0:6600' failed (continuing anyway, because binding to '[::]:6600' succeeded): Failed to bind socket: Address already in use
libsamplerate: libsamplerate converter 'Fastest Sinc Interpolator'
vorbis: Xiph.Org libVorbis 1.3.6
opus: libopus 1.3
sndfile: libsndfile-1.0.28
hybrid_dsd: The Hybrid DSD decoder is disabled because it was not explicitly enabled
wildmidi: configuration file does not exist: /etc/timidity/timidity.cfg
simple_db: reading DB
exception: Input plugin 'tidal' is unavailable: No Tidal application token configured
exception: Input plugin 'qobuz' is unavailable: No Qobuz app_id configured
curl: version 7.62.0
curl: with OpenSSL/1.1.1
avahi: Initializing interface
exception: RTIOThread could not get realtime scheduling, continuing anyway: sched_setscheduler failed: Operation not permitted
avahi: Client changed to state 2
avahi: Client is RUNNING
avahi: Registering service _mpd._tcp/MPD at Raspberry Pi
avahi: Service group changed to state 0
avahi: Service group is UNCOMMITED
state_file: Loading state file /var/lib/mpd/mpdstate
inotify: initializing inotify
decoder_thread: probing plugin mad
decoder: audio_format=44100:24:2, seekable=true
exception: OutputThread could not get realtime scheduling, continuing anyway: sched_setscheduler failed: Operation not permitted
alsa_output: opened hw:0,0 type=HW
alsa_output: buffer: size=64..65536 time=1451..1486078
alsa_output: period: size=32..32768 time=725..743039
alsa_output: default period_time = buffer_time/4 = 500000/4 = 125000
alsa_output: format=S24_LE (Signed 24 bit Little Endian)
alsa_output: buffer_size=22050 period_size=4410
output: opened "Hifiberry" (alsa) audio_format=44100:24:2
inotify: watching music directory
avahi: Service group changed to state 1
avahi: Service group is REGISTERING
client: [0] opened from 192.168.178.201:53866
avahi: Service group changed to state 2
avahi: Service 'MPD at Raspberry Pi' successfully established.
```

**Comment by itai33** [comment]
I'm experiencing this problem as well, using ncmpcpp on arch. I'm unsure what further information to provide, but I'd be happy to provide any if requested.

**Comment by MaxKellermann** [comment]
What was the last MPD version which did not show this behavior? Does it also happen with the built-in DAC? Does it also happen with the "recorder" output plugin, i.e. can you see repeated frames in the recorded file?

**Comment by The-Michael-R** [comment]
* Identifying the exact version that started that behaviour might take some time.
* Using the internal DAC need some time as my box isn't very accessible to plugin something. I'll try.
* To identify the the exact length of the stutter I might be able to measure that with a scope I have flying around here. Allow me also to provide that a bit later as I'm rhight now a bit busy ;-)
* Using the recorder ouput plugin generates a clean record (no stutter).
* Using `ffplay -nodisp <somefile>` plays also fine.

By the way here's my `mpd.conf` if that might help as I have some settings for the output-plugin:
```
# See: /usr/share/doc/mpd/mpdconf.exampley

zeroconf_name           "MPD at Raspberry Pi"

music_directory         "/usr/local/music"
playlist_directory      "/var/lib/mpd/playlists"
pid_file                "/run/mpd/mpd.pid"
db_file                 "/var/lib/mpd/mpd.db"
state_file              "/var/lib/mpd/mpdstate"
#log_file               "/var/log/mpd.log"

bind_to_address         "127.0.0.1"
bind_to_address         "192.168.178.210"

connection_timeout      "2"
auto_update_depth       "3"

user                    "mpd"

audio_output {
        name            "Hifiberry"
        type            "alsa"
        device          "hw:0,0"
        mixer_type      "hardware"
        mixer_control   "DIGITAL"
        mixer_device    "hw:0"
        mixer_index     "0"
        auto_resample   "no"
        auto_format     "no"
        enabled         "yes"
}
```
By the way: Thanks to the MPD team for that great piece of software!

**Comment by MaxKellermann** [comment]
Did you previously have a MPD version which did not have this bug? You wrote "Since the last update of mpd", but what version were you previously running which did not have the bug?
MPD 0.21 had a major change in the ALSA plugin, switching from the blocking API to the non-blocking API, which has serious advantages, but new code may have new bugs, and this may trigger bugs in drivers which may not be visible with other software. So it may be very helpful to know if this also occurs with MPD 0.20.x.
Does it happen after unpausing playback? After seeking?

**Comment by The-Michael-R** [comment]
Yes it worked fine in the past. I might have been a bit more clear: I have the problem since a couple of weeks and the box updates regularly. I hoped the problem will be fixed but as I noticed it isn't reported. Said so: The date 0.20.x->0.21.x seems legit. I tried a rollback to 0.20.21-1 but unfortunately there are other dependencies prohibit that.

It happens: Stop->Play, Pause->Play, Playing->Seek
It not happens: Playing a complete play list (end Title->new Title transition is fine)

**Comment by The-Michael-R** [comment]
OK, I got the scope out:
The mpd plays for about 390ms, silence for 100ms and then it plays. It seems that it does _not_ repeat the first 400ms chunk as I imagined. In the scope picture on the left side is the very first piece of the waveform coming out of the Hifiberry after un-pause a Track.
![stutter1](https://user-images.githubusercontent.com/21987859/48365801-356e8900-e6ac-11e8-9cf0-a617f546031e.png)
I hope that helps to identify which buffer could cause this. If you need more tests I'm happy to help  out as much as I can.

**Comment by MaxKellermann** [comment]
Whoa, what a bug report with good information!

Checking your log:
`alsa_output: buffer_size=22050 period_size=4410`
One "period" is 4410 frames, at 44100Hz this translates to 100ms. This just happens to be the exact duration of your silence pattern. So that's it: there is one period of silence in the buffer.

Your "buffer" has 22050 frames, i.e. there are 5 periods in the buffer. The first 4 periods are ok, and the last one is silence. After that, everything is fine. Only the initial fill of the buffer appears to be bad, and after running cross the buffer boundary, everything is good.

Some more data which could be useful: paste the contents of `/proc/asound/card0/pcm0p/sub0/[sh]w_params` while playing.

Can you somehow check whether the silence replaces the real song data, or whether it is inserted into the stream? i.e. is data missing, or is it just postponed by 100ms?

**Comment by The-Michael-R** [comment]
Thanks :-)

Here the requested data. Out of curiosity I already looked for them a few minutes before. It's seems to run @44.1kHz:
```
> cat /proc/asound/card0/pcm0p/sub0/
hw_params     info          prealloc      prealloc_max  status        sw_params     
[root@MPDpi michael]# cat /proc/asound/card0/pcm0p/sub0/sw_params 
tstamp_mode: NONE
period_step: 1
avail_min: 4410
start_threshold: 17640
stop_threshold: 22050
silence_threshold: 0
silence_size: 0
boundary: 1445068800

> cat /proc/asound/card0/pcm0p/sub0/hw_params 
access: RW_INTERLEAVED
format: S24_LE
subformat: STD
channels: 2
rate: 44100 (44100/1)
period_size: 4410
buffer_size: 22050
```
But I fear your last request might be very tricky. Especially with normal tracks. But let's see if I can craft a special sound-file that allows a recognizable pattern. But that might need a bit more time.

**Comment by The-Michael-R** [comment]
OK, some ugly audio-file later. That's the mpd output:
![mpd-nok](https://user-images.githubusercontent.com/21987859/48370913-ce0c0580-e6ba-11e8-9cbb-89ab67c73c5d.png)
And that's how it should look like (ffplay on the very same system):
![ffplay-ok](https://user-images.githubusercontent.com/21987859/48370947-e11ed580-e6ba-11e8-9d46-92c59c19560e.png)
So I would say the audio output is only stopped for the ~98ms.

**Comment by MaxKellermann** [comment]
Thanks, all of your data was very helpful, and I found the problem:
https://github.com/MusicPlayerDaemon/MPD/blob/57e34823d84502/src/output/plugins/AlsaOutputPlugin.cxx#L654
`ring_buffer = new boost::lockfree::spsc_queue<uint8_t>(period_size * 4);`
The ring buffer is only allocated for 4 periods. So if your hardware buffer has 5 periods, the plugin has at most 4 periods of data, and then needs to fill the remaining buffer with silence.
This problem was hit already half a year ago: https://github.com/MusicPlayerDaemon/MPD/issues/260
And my solution was: https://github.com/MusicPlayerDaemon/MPD/commit/e08598e7e252 - i.e. don't generate silence until the PCM is really started (`start_threshold`).
Now the thing is that your `start_threshold` happens to be the same size as our `ring_buffer`, i.e. 4 periods. The plugin copies those 4 periods from `ring_buffer` to the hardware buffer, draining `ring_buffer` empty, and at the same time, the DAC starts playback due to hitting `start_threshold`. Almost immediately, the plugin will be woken up again because the hardware buffer still isn't full, finding that the player thread hasn't refilled the `ring_buffer` (which takes slightly more time than "immediately").
This is a corner case which was not covered by my previous fix.

**Comment by The-Michael-R** [comment]
Great!

And thank you for this quick support here. Please keep up this good work!

**Comment by The-Michael-R** [comment]
I just want to confirm that the fix works after updating to mpd version 0.21.3.

**Comment by nobrowser** [comment]
Hmm, I started having this exact problem after upgrading from 0.21.5 (from debian buster) to the backport of 0.21.26 (as packaged by the debian mpd maintainers team). Regression, but debian or upstream? I'll report on bugs.debian.org and cross-reference.

**Comment by MaxKellermann** [comment]
@nobrowser write a new and complete bug report.
